### PR TITLE
feat: Split long sentences with overlap

### PR DESCRIPTION
**Pull request by attiks**
This will split too long sentences in multiple ones, using overlap.

Refs: RW-944

### CODE REVIEW

**Review by orakili — CHANGES_REQUESTED** [review]
The code is not working as intended due the `$lines[$index] =[...]` when splitting a sentence which is too long.

File: src/Plugin/ocha_ai/TextSplitter/Sentence.php [hunk 1]
```diff
@@ -38,7 +39,27 @@ public function splitText(string $text, ?int $length = NULL, ?int $overlap = NUL
...
+            while (mb_strlen($sentence) > 0) {
+              $new_sentence = Unicode::truncate($sentence, $length - $overlap, TRUE);
+              $sentences[] = $str_overlap . ' ' . $new_sentence;
+              $str_overlap = trim(strrev(Unicode::truncate(strrev($new_sentence), $overlap, TRUE)));
+              $sentence = trim(mb_substr($sentence, mb_strlen($new_sentence)));
+            }
+          }
+          else {
+            $sentences[] = $sentence;
+          }
```
[the remark this answers is not in the document]
Comment (attiks) [reply src/Plugin/ocha_ai/TextSplitter/Sentence.php]: - All code is moved to TextHelper
- Overlap is optional, default none

File: src/Plugin/ocha_ai/TextSplitter/Token.php [hunk 2]
```diff
@@ -102,4 +118,23 @@ public function buildConfigurationForm(array $form, FormStateInterface $form_sta
     return $form;
   }
 
+  /**
+   * Split a too long line in multiple lines.
+   */
+  protected function splitInLines(string $line, int $length) {
+    // Overlap of 10%.
+    $overlap = (int) round($length / 10, 0);
```
[the remark this answers is not in the document]
Comment (attiks) [reply src/Plugin/ocha_ai/TextSplitter/Token.php]: Removed

File: src/Plugin/ocha_ai/TextSplitter/Token.php [hunk 3]
```diff
@@ -102,4 +118,23 @@ public function buildConfigurationForm(array $form, FormStateInterface $form_sta
...
+    /** @var \Drupal\ocha_ai\Plugin\TextSplitterPluginManager $text_splitter_manager */
+    $text_splitter_manager = \Drupal::service('plugin.manager.ocha_ai.text_splitter');
+
+    /** @var \Drupal\ocha_ai\Plugin\ocha_ai\TextSplitter\Sentence */
+    $splitter = $text_splitter_manager->createInstance('sentence', [
+      'length' => $length,
+      'overlap' => $overlap,
+    ]);
+
+    return $splitter->splitText($line, $length, $overlap);
```
[the remark this answers is not in the document]
Comment (attiks) [reply src/Plugin/ocha_ai/TextSplitter/Token.php]: True, moved

File: src/Plugin/ocha_ai/TextSplitter/Token.php [hunk 4]
```diff
@@ -36,11 +36,27 @@ public function splitText(string $text, ?int $length = NULL, ?int $overlap = NUL
...
       $line_token_count = TextHelper::estimateTokenCount($line);
-      $lines[$index] = [
-        'token_count' => $line_token_count,
-        'text' => $line,
-      ];
-      $total_token_count += $line_token_count;
+
+      // Split a line when it's too long.
+      if ($line_token_count > $max_token_count) {
+        $sub_lines = $this->splitInLines($line, $length);
```
[the remark this answers is not in the document]
Comment (attiks) [reply src/Plugin/ocha_ai/TextSplitter/Token.php]: Method supports now chars or tokens

**Comment by attiks** [comment]
Docs needs some more attention, and tests are needed

File: src/Helpers/TextHelper.php [hunk 5]
```diff
@@ -21,4 +26,271 @@ public static function estimateTokenCount(string $text): int {
...
+      $sentences = self::splitParagraphsInSentences($paragraph, $pattern_id);
+
+      foreach ($sentences as $sentence) {
+        if (self::getLength($sentence, $unit) < $length) {
+          $output[] = $sentence;
+          continue;
+        }
+
+        // Line is too long.
+        while (mb_strlen($sentence) > 0) {
```
Comment (orakili) [inline src/Helpers/TextHelper.php:150]: `$sentence` is an array (see `:: splitParagraphsInSentences()`) so this throws a TypeError exception.

It's missing something like `$sentence = $sentence['text'];` before the `while`.

**Review by orakili — CHANGES_REQUESTED** [review]
The new token splitter doesn't return the results that I would expect.

It doesn't really respect sentence or paragraph boundaries and each chunk ends up containing partial sentences with or without overlap (see example below).

I would expect the splitter to try to aggregate as many paragraphs as possible under the max token count. Then split by sentences if too long and if that's not enough because some sentences are too long then split those long sentences into equal chunks.

With the new splitter, each chunk can contain out of context partial text and I think that would decrease the quality of the chat results notably.

The old token splitter was splitting by paragraphs but was missing sone logic to split too long paragraphs into sentences and too long sentences into smaller text chunks.

**Example**

Given the text below (length: 384, overlap: 0):

```
Mollit quis sunt do dolore irure aliqua ea magna enim aliquip adipisicing esse culpa id elit labore irure ad tempor enim commodo nulla nisi magna fugiat aute exercitation ex sint quis magna incididunt ut proident ad qui commodo incididunt enim anim et pariatur duis labore amet id ipsum exercitation dolor culpa nulla reprehenderit id magna in do commodo pariatur excepteur amet eu ullamco proident labore et irure proident ad laborum aute laborum id ea ex culpa excepteur eu amet aliquip et sint in deserunt quis aliquip labore labore incididunt non quis aliquip proident laboris consequat pariatur non eu dolore deserunt.

Dolore consectetur incididunt magna fugiat ut consequat mollit reprehenderit sit esse dolor non ullamco aliqua nulla irure veniam do exercitation laboris sunt occaecat officia reprehenderit tempor quis irure exercitation labore est consequat dolore commodo sunt est cillum adipisicing elit nulla ex Lorem proident proident nisi culpa esse consectetur irure reprehenderit ipsum do qui esse et velit id do culpa reprehenderit elit nisi officia sit ea non enim occaecat quis excepteur fugiat aliquip sit elit ea veniam sunt labore cillum commodo in ipsum deserunt occaecat nisi pariatur minim eiusmod ipsum quis proident id cillum ullamco consectetur id duis est veniam velit.

Nulla tempor ea reprehenderit dolor veniam est in deserunt eiusmod enim Lorem eu eiusmod et do irure tempor fugiat Lorem qui qui do consectetur ea id laboris dolor nulla sit labore et aliqua quis amet fugiat sunt consectetur qui consectetur pariatur laboris deserunt consectetur occaecat non ex cupidatat nisi aliquip duis laboris amet duis nulla ex fugiat pariatur dolore culpa veniam irure eiusmod dolor excepteur nisi reprehenderit non dolore occaecat labore sunt quis consequat ut reprehenderit magna et velit labore commodo exercitation elit culpa sunt laborum deserunt aute amet quis consectetur enim mollit ullamco tempor adipisicing id proident culpa eiusmod.

Et irure consectetur culpa irure duis reprehenderit nisi Lorem dolor consequat exercitation minim commodo tempor deserunt pariatur Lorem incididunt culpa occaecat laborum dolor ullamco nostrud adipisicing magna qui irure adipisicing est pariatur dolor ea labore ea ad ut eu pariatur ex eiusmod mollit labore duis fugiat anim cupidatat ad esse commodo veniam ad magna ipsum anim velit nostrud consequat anim duis in deserunt ea consectetur laborum nostrud sit veniam proident ex reprehenderit mollit excepteur laboris occaecat esse mollit aliquip nostrud eu occaecat labore pariatur velit aliquip ea sit dolor in non sint occaecat aliquip labore adipisicing enim duis anim nulla id eu laboris exercitation ipsum irure amet anim laboris mollit quis consectetur aliqua est labore incididunt id ullamco reprehenderit cupidatat nulla occaecat esse ut consequat ea Lorem adipisicing ut veniam duis elit irure laboris quis fugiat aliquip veniam tempor incididunt quis aliquip ea ut laborum cupidatat reprehenderit culpa mollit dolor laborum duis Lorem ut qui sunt officia eiusmod veniam dolore velit nisi sit aliqua ad dolor magna veniam excepteur culpa id occaecat anim sunt mollit eu ut esse ex do commodo excepteur esse exercitation eu excepteur eiusmod aliqua quis officia amet pariatur sunt consectetur incididunt officia laborum qui duis ut aute nisi aute est ipsum consequat et ut excepteur ea est mollit consectetur non anim excepteur adipisicing esse quis eiusmod non ea tempor labore non mollit magna mollit laboris eiusmod non aliquip tempor culpa reprehenderit aute consequat id aute sit eu est ut cillum mollit irure in minim dolore nostrud consectetur incididunt mollit minim pariatur esse aliqua proident laboris enim sit culpa nulla dolore adipisicing duis enim dolore aliqua nostrud ipsum cupidatat cillum est in duis duis ullamco dolor veniam nisi labore excepteur enim incididunt eiusmod qui tempor Lorem tempor eiusmod cillum culpa amet anim consequat quis amet.

Quis est ullamco et id labore incididunt esse Lorem sit amet deserunt non labore amet reprehenderit exercitation anim excepteur reprehenderit labore id aute ullamco non adipisicing nisi deserunt tempor aute ipsum ex pariatur enim cupidatat laborum ea consectetur minim excepteur excepteur proident dolore ullamco officia sit tempor laborum ex fugiat deserunt nulla commodo consequat minim consequat elit consectetur qui duis laborum dolor magna anim laboris consectetur adipisicing elit eiusmod cupidatat cupidatat laboris laborum ea culpa cupidatat culpa commodo velit exercitation quis anim laborum consectetur occaecat excepteur aliqua nulla labore ea incididunt anim est irure laborum enim proident pariatur dolore est.

Duis dolor ipsum sunt irure est aliqua aliquip dolore quis cupidatat pariatur qui excepteur fugiat cupidatat et esse id id commodo eiusmod amet sint qui consectetur excepteur fugiat elit aliqua pariatur dolore ut velit ea id consectetur dolor non officia aliqua ullamco adipisicing nostrud dolore ex excepteur sit anim ad enim deserunt nostrud eiusmod minim Lorem eiusmod magna ipsum culpa irure sunt eiusmod quis ex anim culpa nisi quis Lorem veniam minim ullamco deserunt et ut velit enim eiusmod ex nulla in cillum aliquip non reprehenderit veniam proident non nulla do dolore magna aliquip duis elit ipsum eu ipsum duis. Do veniam culpa officia anim dolore ea labore culpa quis qui nisi Lorem ad exercitation consequat est dolor dolor id do pariatur magna nostrud laborum veniam veniam enim nisi fugiat aliquip veniam do aliqua ea laborum nostrud fugiat in veniam cupidatat non irure aliqua Lorem sint nostrud dolor quis velit occaecat ex mollit adipisicing tempor tempor fugiat proident in quis mollit enim officia id incididunt magna adipisicing anim Lorem fugiat adipisicing labore officia dolor magna reprehenderit exercitation adipisicing nisi consequat et proident laborum dolore id esse culpa eu cupidatat cupidatat irure ea cillum eu dolore ut velit esse ex deserunt. Cillum incididunt quis aute in officia aliquip consectetur consectetur id mollit consequat nisi esse pariatur occaecat ipsum Lorem laborum nulla aliquip anim do qui sunt ipsum magna laboris dolore exercitation cillum cupidatat id ipsum laborum elit ullamco Lorem ea do ex sunt eu sint et tempor consequat consectetur officia consectetur esse ipsum elit ea aliqua commodo ipsum officia commodo aliquip ipsum sit ea irure laborum pariatur cupidatat enim officia aliquip id incididunt tempor cillum ipsum consequat ea dolore excepteur id id duis laborum mollit exercitation deserunt non ut eu esse cupidatat anim aliqua sit tempor officia occaecat ipsum veniam qui.
```

The new splitter returns:

```
            [0] => Mollit quis sunt do dolore irure aliqua ea magna enim aliquip adipisicing esse culpa id elit labore irure ad tempor enim commodo nulla nisi magna fugiat aute exercitation ex sint quis magna incididunt ut proident ad qui commodo incididunt enim anim et pariatur duis labore amet id ipsum exercitation dolor culpa nulla reprehenderit id magna in do commodo pariatur excepteur amet eu ullamco proident labore et irure proident ad laborum aute laborum id ea ex culpa excepteur eu amet aliquip et sint in deserunt quis aliquip labore labore incididunt non quis aliquip proident laboris consequat pariatur non eu dolore deserunt. Dolore consectetur incididunt magna fugiat ut consequat mollit reprehenderit sit esse dolor non ullamco aliqua nulla irure veniam do exercitation laboris sunt occaecat officia reprehenderit tempor quis irure exercitation labore est consequat dolore commodo sunt est cillum adipisicing elit nulla ex Lorem proident proident nisi culpa esse consectetur irure reprehenderit ipsum do qui esse et velit id do culpa reprehenderit elit nisi officia sit ea non enim occaecat quis excepteur fugiat aliquip sit elit ea veniam sunt labore cillum commodo in ipsum deserunt occaecat nisi pariatur minim eiusmod ipsum quis proident id cillum ullamco consectetur id duis est veniam velit. Nulla tempor ea reprehenderit dolor veniam est in deserunt eiusmod enim Lorem eu
            [1] => eiusmod et do irure tempor fugiat Lorem qui qui do consectetur ea id laboris dolor nulla sit labore et aliqua quis amet fugiat sunt consectetur qui consectetur pariatur laboris deserunt consectetur occaecat non ex cupidatat nisi aliquip duis laboris amet duis nulla ex fugiat pariatur dolore culpa veniam irure eiusmod dolor excepteur nisi reprehenderit non dolore occaecat labore sunt quis consequat ut reprehenderit magna et velit labore commodo exercitation elit culpa sunt laborum deserunt aute amet quis consectetur enim mollit ullamco tempor adipisicing id proident culpa eiusmod. Et irure consectetur culpa irure duis reprehenderit nisi Lorem dolor consequat exercitation minim commodo tempor deserunt pariatur Lorem incididunt culpa occaecat laborum dolor ullamco nostrud adipisicing magna qui irure adipisicing est pariatur dolor ea labore ea ad ut eu pariatur ex eiusmod mollit labore duis fugiat anim cupidatat ad esse commodo veniam ad magna ipsum anim velit nostrud consequat anim duis in deserunt ea consectetur laborum nostrud sit veniam proident ex reprehenderit mollit excepteur laboris occaecat esse mollit aliquip nostrud eu occaecat labore pariatur velit aliquip ea sit dolor in non sint occaecat aliquip labore adipisicing enim duis anim nulla id eu laboris exercitation ipsum irure amet anim laboris mollit quis consectetur aliqua est labore incididunt id
            [2] => ullamco reprehenderit cupidatat nulla occaecat esse ut consequat ea Lorem adipisicing ut veniam duis elit irure laboris quis fugiat aliquip veniam tempor incididunt quis aliquip ea ut laborum cupidatat reprehenderit culpa mollit dolor laborum duis Lorem ut qui sunt officia eiusmod veniam dolore velit nisi sit aliqua ad dolor magna veniam excepteur culpa id occaecat anim sunt mollit eu ut esse ex do commodo excepteur esse exercitation eu excepteur eiusmod aliqua quis officia amet pariatur sunt consectetur incididunt officia laborum qui duis ut aute nisi aute est ipsum consequat et ut e xcepteur ea est mollit consectetur non anim excepteur adipisicing esse quis eiusmod non ea tempor labore non mollit magna mollit laboris eiusmod non aliquip tempor culpa reprehenderit aute consequat id aute sit eu est ut cillum mollit irure in minim dolore nostrud consectetur incididunt mollit minim pariatur esse aliqua proident laboris enim sit culpa nulla dolore adipisicing duis enim dolore aliqua nostrud ipsum cupidatat cillum est in duis duis ullamco dolor veniam nisi labore excepteur enim incididunt eiusmod qui tempor Lorem tempor eiusmod cillum culpa amet anim consequat quis amet. Quis est ullamco et id labore incididunt esse Lorem sit amet deserunt non labore amet reprehenderit exercitation anim excepteur reprehenderit labore id aute ullamco non adipisicing nisi deserunt tempor aute ipsum ex pariatur enim cupidatat laborum ea consectetur minim excepteur excepteur proident dolore ullamco officia sit tempor laborum ex fugiat deserunt nulla commodo consequat minim consequat elit consectetur qui duis laborum dolor magna anim laboris consectetur adipisicing elit eiusmod cupidatat cupidatat laboris laborum ea culpa cupidatat culpa commodo velit exercitation quis anim laborum consectetur occaecat excepteur aliqua nulla labore ea incididunt anim est irure laborum enim proident pariatur dolore est.
            [3] => Duis dolor ipsum sunt irure est aliqua aliquip dolore quis cupidatat pariatur qui excepteur fugiat cupidatat et esse id id commodo eiusmod amet sint qui consectetur excepteur fugiat elit aliqua pariatur dolore ut velit ea id consectetur dolor non officia aliqua ullamco adipisicing nostrud dolore ex excepteur sit anim ad enim deserunt nostrud eiusmod minim Lorem eiusmod magna ipsum culpa irure sunt eiusmod quis ex anim culpa nisi quis Lorem veniam minim ullamco deserunt et ut velit enim eiusmod ex nulla in cillum aliquip non reprehenderit veniam proident non nulla do dolore magna aliquip duis elit ipsum eu ipsum duis. Do veniam culpa officia anim dolore ea labore culpa quis qui nisi Lorem ad exercitation consequat est dolor dolor id do pariatur magna nostrud laborum veniam veniam enim nisi fugiat aliquip veniam do aliqua ea laborum nostrud fugiat in veniam cupidatat non irure aliqua Lorem sint nostrud dolor quis velit occaecat ex mollit adipisicing tempor tempor fugiat proident in quis mollit enim officia id incididunt magna adipisicing anim Lorem fugiat adipisicing labore officia dolor magna reprehenderit exercitation adipisicing nisi consequat et proident laborum dolore id esse culpa eu cupidatat cupidatat irure ea cillum eu dolore ut velit esse ex deserunt. Cillum incididunt quis aute in officia aliquip consectetur consectetur id mollit consequat nisi esse
            [4] =>  pariatur occaecat ipsum Lorem laborum nulla aliquip anim do qui sunt ipsum magna laboris dolore exercitation cillum cupidatat id ipsum laborum elit ullamco Lorem ea do ex sunt eu sint et tempor consequat consectetur officia consectetur esse ipsum elit ea aliqua commodo ipsum officia commodo aliquip ipsum sit ea irure laborum pariatur cupidatat enim officia aliquip id incididunt tempor cillum ipsum consequat ea dolore excepteur id id duis laborum mollit exercitation deserunt non ut eu esse cupidatat anim aliqua sit tempor officia occaecat ipsum veniam qui.
```

While I would have expected something like:

```
            [0] => Mollit quis sunt do dolore irure aliqua ea magna enim aliquip adipisicing esse culpa id elit labore irure ad tempor enim commodo nulla nisi magna fugiat aute exercitation ex sint quis magna incididunt ut proident ad qui commodo incididunt enim anim et pariatur duis labore amet id ipsum exercitation dolor culpa nulla reprehenderit id magna in do commodo pariatur excepteur amet eu ullamco proident labore et irure proident ad laborum aute laborum id ea ex culpa excepteur eu amet aliquip et sint in deserunt quis aliquip labore labore incididunt non quis aliquip proident laboris consequat pariatur non eu dolore deserunt.

Dolore consectetur incididunt magna fugiat ut consequat mollit reprehenderit sit esse dolor non ullamco aliqua nulla irure veniam do exercitation laboris sunt occaecat officia reprehenderit tempor quis irure exercitation labore est consequat dolore commodo sunt est cillum adipisicing elit nulla ex Lorem proident proident nisi culpa esse consectetur irure reprehenderit ipsum do qui esse et velit id do culpa reprehenderit elit nisi officia sit ea non enim occaecat quis excepteur fugiat aliquip sit elit ea veniam sunt labore cillum commodo in ipsum deserunt occaecat nisi pariatur minim eiusmod ipsum quis proident id cillum ullamco consectetur id duis est veniam velit.
            [1] => Nulla tempor ea reprehenderit dolor veniam est in deserunt eiusmod enim Lorem eu eiusmod et do irure tempor fugiat Lorem qui qui do consectetur ea id laboris dolor nulla sit labore et aliqua quis amet fugiat sunt consectetur qui consectetur pariatur laboris deserunt consectetur occaecat non ex cupidatat nisi aliquip duis laboris amet duis nulla ex fugiat pariatur dolore culpa veniam irure eiusmod dolor excepteur nisi reprehenderit non dolore occaecat labore sunt quis consequat ut reprehenderit magna et velit labore commodo exercitation elit culpa sunt laborum deserunt aute amet quis consectetur enim mollit ullamco tempor adipisicing id proident culpa eiusmod.
            [2] => Et irure consectetur culpa irure duis reprehenderit nisi Lorem dolor consequat exercitation minim commodo tempor deserunt pariatur Lorem incididunt culpa occaecat laborum dolor ullamco nostrud adipisicing magna qui irure adipisicing est pariatur dolor ea labore ea ad ut eu pariatur ex eiusmod mollit labore duis fugiat anim cupidatat ad esse commodo veniam ad magna ipsum anim velit nostrud consequat anim duis in deserunt ea consectetur laborum nostrud sit veniam proident ex reprehenderit mollit excepteur laboris occaecat esse mollit aliquip nostrud eu occaecat labore pariatur velit aliquip ea sit dolor in non sint occaecat aliquip labore adipisicing enim duis anim nulla id eu laboris exercitation ipsum irure amet anim laboris mollit quis consectetur aliqua est labore incididunt id ullamco reprehenderit cupidatat nulla occaecat esse ut consequat ea Lorem adipisicing ut veniam duis elit irure laboris quis fugiat aliquip veniam tempor incididunt quis aliquip ea ut laborum cupidatat reprehenderit culpa mollit dolor laborum duis Lorem ut qui sunt officia eiusmod veniam dolore velit nisi sit aliqua ad dolor magna veniam excepteur 
            [3] culpa id occaecat anim sunt mollit eu ut esse ex do commodo excepteur esse exercitation eu excepteur eiusmod aliqua quis officia amet pariatur sunt consectetur incididunt officia laborum qui duis ut aute nisi aute est ipsum consequat et ut excepteur ea est mollit consectetur non anim excepteur adipisicing esse quis eiusmod non ea tempor labore non mollit magna mollit laboris eiusmod non aliquip tempor culpa reprehenderit aute consequat id aute sit eu est ut cillum mollit irure in minim dolore nostrud consectetur incididunt mollit minim pariatur esse aliqua proident laboris enim sit culpa nulla dolore adipisicing duis enim dolore aliqua nostrud ipsum cupidatat cillum est in duis duis ullamco dolor veniam nisi labore excepteur enim incididunt eiusmod qui tempor Lorem tempor eiusmod cillum culpa amet anim consequat quis amet.
            [4] => Quis est ullamco et id labore incididunt esse Lorem sit amet deserunt non labore amet reprehenderit exercitation anim excepteur reprehenderit labore id aute ullamco non adipisicing nisi deserunt tempor aute ipsum ex pariatur enim cupidatat laborum ea consectetur minim excepteur excepteur proident dolore ullamco officia sit tempor laborum ex fugiat deserunt nulla commodo consequat minim consequat elit consectetur qui duis laborum dolor magna anim laboris consectetur adipisicing elit eiusmod cupidatat cupidatat laboris laborum ea culpa cupidatat culpa commodo velit exercitation quis anim laborum consectetur occaecat excepteur aliqua nulla labore ea incididunt anim est irure laborum enim proident pariatur dolore est.
            [5] => Duis dolor ipsum sunt irure est aliqua aliquip dolore quis cupidatat pariatur qui excepteur fugiat cupidatat et esse id id commodo eiusmod amet sint qui consectetur excepteur fugiat elit aliqua pariatur dolore ut velit ea id consectetur dolor non officia aliqua ullamco adipisicing nostrud dolore ex excepteur sit anim ad enim deserunt nostrud eiusmod minim Lorem eiusmod magna ipsum culpa irure sunt eiusmod quis ex anim culpa nisi quis Lorem veniam minim ullamco deserunt et ut velit enim eiusmod ex nulla in cillum aliquip non reprehenderit veniam proident non nulla do dolore magna aliquip duis elit ipsum eu ipsum duis. Do veniam culpa officia anim dolore ea labore culpa quis qui nisi Lorem ad exercitation consequat est dolor dolor id do pariatur magna nostrud laborum veniam veniam enim nisi fugiat aliquip veniam do aliqua ea laborum nostrud fugiat in veniam cupidatat non irure aliqua Lorem sint nostrud dolor quis velit occaecat ex mollit adipisicing tempor tempor fugiat proident in quis mollit enim officia id incididunt magna adipisicing anim Lorem fugiat adipisicing labore officia dolor magna reprehenderit exercitation adipisicing nisi consequat et proident laborum dolore id esse culpa eu cupidatat cupidatat irure ea cillum eu dolore ut velit esse ex deserunt. 
            [6] Cillum incididunt quis aute in officia aliquip consectetur consectetur id mollit consequat nisi esse pariatur occaecat ipsum Lorem laborum nulla aliquip anim do qui sunt ipsum magna laboris dolore exercitation cillum cupidatat id ipsum laborum elit ullamco Lorem ea do ex sunt eu sint et tempor consequat consectetur officia consectetur esse ipsum elit ea aliqua commodo ipsum officia commodo aliquip ipsum sit ea irure laborum pariatur cupidatat enim officia aliquip id incididunt tempor cillum ipsum consequat ea dolore excepteur id id duis laborum mollit exercitation deserunt non ut eu esse cupidatat anim aliqua sit tempor officia occaecat ipsum veniam qui. 
```

File: src/Helpers/TextHelper.php (hunk 5 above)
Comment (attiks) [reply src/Helpers/TextHelper.php]: Code rewritten

**Review by orakili — APPROVED** [review]
Thanks. That's closer to the expected results.

The paragraph separations are not preserved:

```
Paragraph1.

Paragraph2.
```

becomes 

```
Paragraph1. Paragraph2.
```

I'm not sure if that has a huge impact. In my tests, newer models tend to notice paragraphs and are able to work with separate paragraphs (ex: summarize each paragraph) while older ones not so much.

For our use cases, maybe it doesn't matter much.

**Comment by attiks** [comment]
Made a small - unrelated - change to keep the sentences of a paragraph together if possible.

I'll create a a ticket to preserve paragraphs

**Pull request merged**